### PR TITLE
Adding Text Analytics module (Cognitive Services)

### PR DESCRIPTION
**Pull request by neilpeterson**
@krancour coming back around to this. This PR adds the Cognitive Services Text Analytics API to OSBA. I had initially submitted a related PR many months back, however that was delayed due to an ongoing refactor of the broker.

Here is the initial PR - https://github.com/Azure/open-service-broker-azure/pull/463

I've refactored to align with the current state of the broker.

One recommendation in the initial PR was to shrink this down into a single plan and specify the tier with a provisioning parameter. I am cool with this, but have not done so yet. One of plan properties is `Free`. Text Analytics has a free tier, and then several others. If using a single plan however would you suggest I use the `Free` property.

Thanks a bunch for checking this out. 

## Tests

Here are the results of `make test-service-lifecycles`:

```
2018/08/23 23:44:52 ----> creating resource group "test-6652437c-e49b-47f3-943d-b6eaafe0e640"
2018/08/23 23:44:55 ----> created resource group "test-6652437c-e49b-47f3-943d-b6eaafe0e640"
=== RUN   TestServices
2018/08/23 23:44:55 ----> testing in resource group "test-6652437c-e49b-47f3-943d-b6eaafe0e640"
2018/08/23 23:44:55 Running tests for modules: [textanalytics]
=== RUN   TestServices/lifecycle
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-free
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-free
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s0
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s0
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s1
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s1
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s2
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s2
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s3
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s3
=== RUN   TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s4
=== PAUSE TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s4
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-free
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-free: starting
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s4
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s4: starting
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s3
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s3: starting
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s2
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s2: starting
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s1
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s1: starting
=== CONT  TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s0
2018/08/23 23:44:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s0: starting
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s3: in progress
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s4: in progress
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s1: in progress
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s2: in progress
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-free: in progress
2018/08/23 23:45:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s0: in progress
2018/08/23 23:46:31 ----> TestServices/lifecycle/textanalytics/text-analytics-s1: completed
2018/08/23 23:46:32 ----> TestServices/lifecycle/textanalytics/text-analytics-free: completed
2018/08/23 23:46:32 ----> TestServices/lifecycle/textanalytics/text-analytics-s0: completed
2018/08/23 23:46:33 ----> TestServices/lifecycle/textanalytics/text-analytics-s3: completed
2018/08/23 23:46:46 ----> TestServices/lifecycle/textanalytics/text-analytics-s4: completed
2018/08/23 23:46:55 ----> TestServices/lifecycle/textanalytics/text-analytics-s2: in progress
2018/08/23 23:47:14 ----> TestServices/lifecycle/textanalytics/text-analytics-s2: completed
--- PASS: TestServices (139.27s)
    --- PASS: TestServices/lifecycle (0.00s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s1 (96.44s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-free (96.87s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s0 (97.35s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s3 (98.58s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s4 (111.62s)
        --- PASS: TestServices/lifecycle/TestServices/lifecycle/textanalytics/text-analytics-s2 (139.26s)
PASS
2018/08/23 23:47:14 ----> deleting resource group "test-6652437c-e49b-47f3-943d-b6eaafe0e640"
2018/08/23 23:47:16 ----> deleted resource group "test-6652437c-e49b-47f3-943d-b6eaafe0e640"
ok      github.com/Azure/open-service-broker-azure/tests/lifecycle      143.266s
```

### CODE REVIEW

**Comment by jeremyrickard** [comment]
@neilpeterson I've gone through this and I like the addition, thanks!

While running CI, it seems there is an issue w/ the vendored code. Could you ensure that the dep config/vendor is updated?

**Comment by neilpeterson** [comment]
@jeremyrickard - thanks. I've run `make dep` which seems to have cleared up the issue.

I'm not super familiar with package management in Go, if I've done something incorrect here, I apologize :). If so, can you provide a pointer?

Thanks

File: tests/lifecycle/textanalytics_cases_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+var textanalyticsTestCases = []serviceLifecycleTestCase{
+	{ // Text analytics free tier
+		group:     "textanalytics",
+		name:      "text-analytics-free",
+		serviceID: "8f6c848a-4ce1-4a69-9248-63545d3e7e9c",
+		planID:    "d5a0f91f-10da-42fc-b792-656a616d9ec2",
+		provisioningParameters: map[string]interface{}{
+			"location": "eastus",
+		},
```
Comment (zhongyi-zhang) [inline tests/lifecycle/textanalytics_cases_test.go]: Could you also add "testCredentials" in the lifecycle cases? It is like an example how the credentials can be used. You can refer to MySQL: https://github.com/Azure/open-service-broker-azure/blob/master/tests/lifecycle/mysql_cases_test.go#L43.

File: pkg/services/textanalytics/plan_schemas.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+				Description: "The Azure region in which to provision" +
+					" applicable resources.",
+				CustomPropertyValidator: azure.LocationValidator,
+			},
+			"resourceGroup": &service.StringPropertySchema{
+				Title: "Resource group",
+				Description: "The (new or existing) resource group with which" +
+					" to associate new resources.",
+			},
+			"name": &service.StringPropertySchema{
```
Comment (norshtein) [inline pkg/services/textanalytics/plan_schemas.go]: Why make parameter `name` user-specified? In other modules, the name of the server, database, and so on, is generated randomly and is not configurable for users. I'm not familiar with textanalytics, please tell me if there is any special concern to do this. If not, could you make `name` randomly generated just like other modules?

File: tests/lifecycle/textanalytics_cases_test.go (hunk 1 above)
Comment (neilpeterson) [reply tests/lifecycle/textanalytics_cases_test.go]: @zhongyi-zhang - I've added a test to validate a response code of 200 from the Text Analytics endpoint using the returned endpoint address and access key. This should be sufficient.

https://westus.dev.cognitive.microsoft.com/docs/services/TextAnalytics.V2.0/operations/56f30ceeeda5650db055a3c7

File: pkg/services/textanalytics/plan_schemas.go (hunk 2 above)
Comment (neilpeterson) [reply pkg/services/textanalytics/plan_schemas.go]: Thanks for catching this @norshtein . The parameter was actually unused and has been removed.

**Comment by neilpeterson** [comment]
@jeremyrickard @norshtein @zhongyi-zhang I've address all noted issues. If you need me to do anything else, please let me know.

Thanks

File: tests/lifecycle/textanalytics_cases_test.go (hunk 1 above)
Comment (zhongyi-zhang) [reply tests/lifecycle/textanalytics_cases_test.go]: @neilpeterson cool, thanks for adding it!

File: tests/lifecycle/textanalytics_cases_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+// Sentiment struct for credentials test
+type Sentiment struct {
+	Documents []Documents `json:"documents"`
+}
+
+// Documents struct for credentials test
+type Documents struct {
+	Language string `json:"language"`
```
Comment (zhongyi-zhang) [inline tests/lifecycle/textanalytics_cases_test.go]: From the doc, is the `Language` useless? Also, it is a little confusing to me to pass the language when detecting language...

Comment (neilpeterson) [reply tests/lifecycle/textanalytics_cases_test.go]: I apologize, I added an incorrect doc. This test returns a sentiment score.

Here is the correct doc.

https://westus.dev.cognitive.microsoft.com/docs/services/TextAnalytics.V2.0/operations/56f30ceeeda5650db055a3c9

**Comment by norshtein** [comment]
Since new merged PR has modified the Gopkg.lock, I'm afraid you have to run `make dep` again to solve the vendored code issue, sorry for inconvenience. Also, please have a look on lint issue, thanks!

**Comment by neilpeterson** [comment]
Thanks @norshtein - I've run `make dep` and have also fixed up the lint issues.

**Comment by zhongyi-zhang** [comment]
@neilpeterson thanks for the contribution! This service is a good first experimental service of that a bunch of Cognitive Services in OSBA.

**Comment by neilpeterson** [comment]
@zhongyi-zhang - thanks a bunch for your help on this one.

**Pull request merged**